### PR TITLE
fix!: Ban Fields in for loop indices and bitwise ops

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/3639
Resolves https://github.com/noir-lang/noir/issues/4193

## Summary\*

Uses the new TypeVariableKind::Integer in for loops and bitwise operations to prevent `Field` types from being used there.

Removes the old `delayed_type_checks` hack.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
Note the PR body is repeated, not sure if you meant to state that it resolves two different issues:

```
Resolves https://github.com/noir-lang/noir/issues/3639
Resolves https://github.com/noir-lang/noir/issues/3639
```

**Comment by jfecher** [comment]
@kevaundray those issues are correct, the other PR didn't quite resolve them since it implemented the new type variable kind but the new kind was unused until this PR

**Comment by sirasistant** [comment]
There is a typing issue in sort_via in std, I commented sort_via and run the println in ssa_gen/codegen_for:
```rust
      println!(
            "Type of start index is {:?} type of iterator is {:?} type of end index is {:?}",
            self.builder.type_of_value(start_index),
            index_type,
            self.builder.type_of_value(end_index)
        );
```
And it still gave:
`Type of start index is Numeric(NativeField) type of iterator is Numeric(Unsigned { bit_size: 32 }) type of end index is Numeric(NativeField)`
in https://github.com/noir-lang/noir/pull/4290 example

**Comment by jfecher** [comment]
@sirasistant good catch, I found a bug when unifying TypeVariableKind::Integer where we still succeeded previously on Fields: https://github.com/noir-lang/noir/pull/4376/commits/6bc4e2364019c23fde88eb6c50f67713bd33bd17

**Comment by jfecher** [comment]
Wow, ok. So after the change it turns out we were using Fields in for loops in a lot of places. So this is actually quite the breaking change:
```
error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:21:15
   │
21 │     for _r in 0..rf / 2 {
   │               ---------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:22:18
   │
22 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:25:18
   │
25 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:33:15
   │
33 │     for _r in 0..rp {
   │               -----
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:34:18
   │
34 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:43:15
   │
43 │     for _r in 0..rf / 2 {
   │               ---------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:44:18
   │
44 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:47:18
   │
47 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:70:14
   │
70 │     for k in 0..msg.len() {
   │              ------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon/bn254.nr:96:14
   │
96 │     for i in 0..input.len() {
   │              --------------
   │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:105:14
    │
105 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:114:14
    │
114 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:123:14
    │
123 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:132:14
    │
132 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:141:14
    │
141 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:150:14
    │
150 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:159:14
    │
159 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:168:14
    │
168 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:177:14
    │
177 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:186:14
    │
186 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:195:14
    │
195 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:204:14
    │
204 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:213:14
    │
213 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:222:14
    │
222 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/poseidon/bn254.nr:231:14
    │
231 │     for i in 0..input.len() {
    │              --------------
    │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:37:14
   │
37 │     for r in 0..(ark.len() / state.len()) {
   │              ---------------------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:38:18
   │
38 │         for i in 0..state.len() {
   │                  --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:44:22
   │
44 │             for i in 1..state.len() {
   │                      --------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:67:14
   │
67 │     for k in 0..msg.len() {
   │              ------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:94:14
   │
94 │     for i in 0..x.len() {
   │              ----------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/poseidon.nr:96:18
   │
96 │         for j in 0..x.len() {
   │                  ----------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/hash/mimc.nr:11:14
   │
11 │     for i in 1..constants.len() {
   │              ------------------
   │

error: The type Field cannot be used in a for loop
    ┌─ std/hash/mimc.nr:120:17
    │
120 │     for elem in array {
    │                 -----
    │

error: The type Field cannot be used in a for loop
    ┌─ std/array.nr:122:14
    │
122 │     for i in 0..a.len() {
    │              ----------
    │

error: The type Field cannot be used in a for loop
  ┌─ std/merkle.nr:9:14
  │
9 │     for i in 0..n {
  │              ----
  │

error: The type Field cannot be used in a for loop
   ┌─ std/sha512.nr:95:14
   │
95 │     for k in 0..msg.len() {
   │              ------------
   │

error: The type Field cannot be used in a for loop
    ┌─ std/ec.nr:164:14
    │
164 │     for i in 0..N_BITS {
    │              ---------
    │

error: Expected type Field, found type u64
   ┌─ std/array.nr:18:48
   │
18 │             let pos = find_index(sorted_index, i);
   │                                                -
   │

error: Types in a binary operation should match, but found Field and u64
   ┌─ std/array.nr:19:20
   │
19 │             assert(sorted_index[pos] == i);
   │                    ----------------------
   │

error: expected type [Field; N], found type [u64; N]
   ┌─ std/array.nr:34:87
   │
34 │     unconstrained fn get_sorting_index<Env>(self, ordering: fn[Env](T, T) -> bool) -> [Field; N] {
   │                                                                                       ---------- expected [Field; N] because of return type
   ·
52 │         result
   │         ------ [u64; N] returned here
   │

error: The type Field cannot be used in a for loop
   ┌─ std/array.nr:59:21
   │
59 │         for elem in self {
   │                     ----
   │

error: The type Field cannot be used in a for loop
   ┌─ std/array.nr:71:18
   │
71 │         for i in 1 .. self.len() {
   │                  ---------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/array.nr:82:21
   │
82 │         for elem in self {
   │                     ----
   │

error: The type Field cannot be used in a for loop
   ┌─ std/array.nr:93:18
   │
93 │         for i in 1 .. self.len() {
   │                  ---------------
   │

error: The type Field cannot be used in a for loop
    ┌─ std/array.nr:102:21
    │
102 │         for elem in self {
    │                     ----
    │

error: The type Field cannot be used in a for loop
    ┌─ std/array.nr:111:21
    │
111 │         for elem in self {
    │                     ----
    │

error: The type Field cannot be used in a for loop
   ┌─ std/collections/bounded_vec.nr:47:18
   │
47 │         for i in 0..array.len() {
   │                  --------------
   │

error: Types in a binary operation should match, but found Field and u64
   ┌─ std/collections/bounded_vec.nr:60:29
   │
60 │             exceeded_len |= i == append_len;
   │                             ---------------
   │

error: Types in a binary operation should match, but found Field and u64
   ┌─ std/collections/bounded_vec.nr:81:29
   │
81 │             exceeded_len |= i == self.len;
   │                             -------------
   │

error: The type Field cannot be used in a for loop
   ┌─ std/slice.nr:38:21
   │
38 │         for elem in other {
   │                     -----
   │

error: The type Field cannot be used in a for loop
   ┌─ std/cmp.nr:24:18
   │
24 │         for i in 0 .. self.len() {
   │                  ---------------
   │

error: The type Field cannot be used in a for loop
    ┌─ std/cmp.nr:201:18
    │
201 │         for i in 0 .. self.len() {
    │                  ---------------
    │
````

**Comment by jfecher** [comment]
@sirasistant I've found the core issue in #4290 is a result from a bug in the monomorphizer. I've pulled out the fix here: https://github.com/noir-lang/noir/pull/4386. In the general case though, this PR will still be required to prevent Fields in for loops.

**Comment by Savio-Sou** [comment]
cc @noir-lang/developerrelations we have put up warnings for this breaking change for quite some time now, but flagging for early heads up as this might still break quite some code from the community once merged and released.

**Comment by sirasistant** [comment]
Should we make len() return type be an usize (u64)?

**Comment by jfecher** [comment]
@sirasistant I made that change in this PR but I'm open to debate on whether we should keep it or not.

**Comment by sirasistant** [comment]
Ah cool, I haven't look at the code yet. That sounds good to me :D

File: noir_stdlib/src/collections/bounded_vec.nr [hunk 1]
```diff
@@ -12,7 +10,7 @@ impl<T, MaxLen> BoundedVec<T, MaxLen> {
     }
 
     pub fn get(mut self: Self, index: Field) -> T {
```
Comment (kevaundray) [inline noir_stdlib/src/collections/bounded_vec.nr:12]: Might make sense for this be a u64 instead of a Field

**Review by kevaundray — APPROVED** [review]

**Comment by TomAFrench** [comment]
Spoke in person but for posterity, this is because we ignore the lockfile in the "MSRV check" workflow as this ensures that we'll be able to publish to crates.io because we want to know if this will fail before we push out a new version.

We shouldn't block non-release PRs on this however. I'll update this to run on master and raise issues in future.

**Pull request merged**